jax/_src/lax/slicing.py

```diff
@@ -1411,10 +1381,10 @@ def _gather_lower(ctx, operand, indices, *,
                   indices_are_sorted, mode, fill_value):
   aval_out, = ctx.avals_out
   if core.is_opaque_dtype(aval_out.dtype):
-    return aval_out.dtype._rules.gather_mlir(
-        ctx, operand, indices, dimension_numbers=dimension_numbers,
+    return [aval_out.dtype._rules.gather_mlir(
+        ctx, ctx.avals_in, aval_out, operand, indices, dimension_numbers=dimension_numbers,
```

Comment:
Minor nitpick, but at least to explain our previous selves: I believe we were keeping a convention that the opaque dtype definition's lowering rule has the same signature as the primitive lowering rule (so `FooTy._rules.gather_mlir` didn't accept the in/out avals because `_gather_lower` here doesn't either). We don't have to stick to that. Any consistent convention is fine and I like this one as well, where the opaque dtype's rules consistently accept in/out avals as arguments.

Reply:
Understood. I think that this one is more flexible because you can then use, e.g., `broadcast_in_dim_mlir` in the lowering of some other primitive, e.g., `svd`.